proposals/0072-feature-gate-threshold-automation.md

```diff
@@ -0,0 +1,261 @@
+except `E` and `H`, their `u8` value would be 246, or `11110110`.
+
+The `SignalSupportForStagedFeatures` instruction processor will provide the
+vote account's address to the `GetEpochStake` syscall to retrieve the stake
+delegated to that vote account for the epoch. Then, using the `1` values
+provided in the bitmask, the processor will add this stake value to each
+corresponding feature ID's stake support in the Staged Features PDA.
+
+Nodes should submit a transaction containing this instruction:
+
+- Any time at least 4500 slots before the end of the epoch.
+- During startup after any reboot.
+
+Transactions sent too late (< 4500 slots before the epoch end) will be rejected
+by the Feature Gate program.
+
+If a node does not send this transaction or it is rejected, their stake is not
+tallied. This is analogous to a node sending this transaction at a valid slot
+in the epoch signalling support for zero features.
+
+If a feature is revoked, the list of staged features will not change, and nodes
+may still signal support for this feature. However, the runtime will not
+activate this feature if its corresponding feature account no longer exists
+on-chain.
+
+### Step 4: Feature Activation
+
+During the epoch rollover, the runtime must load the Staged Features PDA
+and calculate the stake - as a percentage of the total epoch stake - in support
+for each feature ID to determine which staged features to activate.
+
+Every feature whose stake support meets the required threshold must be
+activated. This threshold will be hard-coded in the runtime to 95% initially,
+but future iterations on the process could allow feature key-holders to set a
+custom threshold per-feature.
+
+As mentioned previously, if a feature was revoked, it will no longer exist
+on-chain, and therefore will be not activated by the runtime, regardless of
+calculated stake support.
+
+If a feature is not activated, either because it has been revoked or it did not
+meet the required stake support, it must be resubmitted according to Step 2.
+
+Once the epoch rollover is complete, the Feature Gate program will reset the
+Staged Features PDA for the new epoch the very first time it is invoked
+without error in the new epoch. Both instructions - `StageFeatureForActivation`
+and `SignalSupportForStagedFeature` - will check the value for `next_epoch`
+against the *current epoch* value in the `Clock` sysvar before executing the
+steps outlined previously in this proposal. If a match is detected, the account
+state is reset for the new epoch as follows:
+
+1. The `current_feature_stakes` list is reset to an empty list.
+2. The `current_epoch` is set to the value of `next_epoch`.
+3. The feature IDs from the `next_features` list are written into the
+   `current_feature_stakes` list with stake support initialized to zero.
+4. The `next_epoch` value is set to the *next* upcoming epoch number
+   (`current_epoch` + 1).
+5. The `next_features` list is reset to an empty list.
```

Comment:
Thanks, this is more clear. However, a question about the design: did you choose this "crank" approach so that the account update would be in the BPF program instead of a runtime responsibility?
I'm a little concerned about edge cases. It seems a lot easier to reason about current and next epoch (and related feature lists) when a reset occurs invariably on every epoch boundary. Otherwise, we foist additional logic on the validators.

For instance, validators need to query the `current_feature` list to determine their support data; but if the list may or may not have been updated yet, they have to look at both epochs and handle the logic correctly. In a similar vein, if there are no features staged for activation, there's no reason for validators to pay for a support-signal transaction at all.

What's the FD take on this @topointon-jump ?

Reply:
> did you choose this "crank" approach so that the account update would be in the BPF program instead of a runtime responsibility?

Yes, that's correct. @topointon-jump and I were leaning toward moving as much responsibility out of the runtime and into the program as possible. I was also aiming for zero garbage collection necessary.

I think a relationship where the runtime clears or otherwise updates account state, which the program has to then depend on, is brittle and should be avoided with these Core BPF programs wherever possible, especially across multiple clients.

I did entertain the idea of having one PDA that gets reused per-epoch to stage features, while using a PDA derived from each epoch to house the feature IDs validators must signal for, but I couldn't think of a good way to "crank" the transition from staged features to signal-ready features without the runtime's help.

Reply:
@CriesofCarrots Yes exactly, the idea was to move as much complexity out of the runtime as possible so that it's easier to write new clients in the future.

Deciding whether or not to use a crank is a trade-off between the complexity introduced by having to deal with the un-cranked state, and the complexity introduced by pushing the account update and garbage collection logic inside the runtime.

We felt that using a crank would reduce the overall complexity on the runtime, but open to other thoughts 😄 do you think differently?

> In a similar vein, if there are no features staged for activation, there's no reason for validators to pay for a support-signal transaction at all.
I would say let's not worry about this case, it's far more likely for there to be staged features than not.

Reply:
> We felt that using a crank would reduce the overall complexity on the runtime, but open to other thoughts 😄 do you think differently?

Yeah, I do think differently. I appreciate the desire to keep as much functionality in the program as possible, but the Feature Gate program is already a special case with regard to runtime responsibility. I think that having accounts the runtime depends on for activations possibly out of sync with the epoch rollover (because they haven't been cranked) is a far more brittle proposition than putting a small extra runtime requirement on clients.

> it's far more likely for there to be staged features than not.

I don't think this is true. Even during the craziest days of the Solana Labs client, the majority of epochs passed without a feature activation. Activations have slowed since then, due to greater care and scheduling around weekends. If anything, I would expect them to happen even less frequently as we move toward greater involvement by multiple parties and cluster governance.

How about we reconsider my original idea:

Use a new PDA for each epoch. The Feature Gate program can use the Clock sysvar to determine the current epoch and allow: (1) staging new features in the PDA for the current epoch; (2) signalling/tallying support for features in the PDA for the previous epoch.
Honestly, I don't think these PDAs ever need to be garbage collected; the number is very small, and they might be kind of nice for historical record purposes, especially if we allow variance of the minimum stake threshold for activation at some point in the future. But if we do feel strongly about them being garbage collected, we could add an unpermissioned instruction to close them that only executes successfully after the signalling/activation epoch has passed.

Seems like this satisfies your desire to keep logic in the program, it satisfies my desire to have runtime-dependency accounts in a known state, and it better satisfies my desire to not foist complex logic on the validators building signal transactions -- they would have to use the epoch number to derive the PDA address, but validators always know what epoch it is anyway.

I think this idea was set aside due to spam concerns, but seems like that isn't an issue with the existing design, since `StageFeatureForActivation` is gated.

Reply:
@buffalojoec ^ Does this work for you?

Reply:
Yeah this works for me. As mentioned previously we no longer need to walk any accounts, so garbage collection is less of a concern.

> (1) staging new features in the PDA for the current epoch; (2) signalling/tallying support for features in the PDA for the previous epoch.

To be totally clear, this is one single PDA with two epochs of information (stage, signals), or two separate PDAs? Either is fine for me, since they're quite small, but two separate could be less brittle.

> But if we do feel strongly about them being garbage collected, we could add an unpermissioned instruction to close them that only executes successfully after the signalling/activation epoch has passed.

Yeah, nice idea. I'm fine with tabling this as a future revision when we may deem necessary, excluding it from this SIMD.

Reply:
> To be totally clear, this is one single PDA with two epochs of information (stage, signals), or two separate PDAs? Either is fine for me, since they're quite small, but two separate could be less brittle.

I think we only need one PDA per epoch, with one epoch's-worth of information. (So I guess two, as you've phrased the question?) Let the epoch that's used in the PDA's address seed equal `N`. That PDA would be available to add (stage) new features during epoch `N`, and available to tally stake support during epoch `N+1`.

Or if we prefer the address seed to align with the signaling epoch (for validator signalling simplicity), we can stage in epoch `N-1` and signal in epoch `N`. I kind of like this better, now that I've typed it out.

Reply:
Yes I like the latter better as well. Stage in `N-1`, signal in `N`.

Reply:
I think the only weird part here is the creation of the PDA when it doesn't exist yet. The multisig "stager" would have to first fund this account if it doesn't exist. This can be implemented CLI/client-side, and the stager would pay the rent lamports on first feature staged, for every epoch.

Reply:
I was envisioning that the optional PDA funding/creation would be part of the program, similar to how the ATA program works.

Reply:
But yes, the first feature staged in an epoch would cost a little more. I think that's fine.

Reply:
Alrighty, let's see if I made it clear enough!